src/rideables/TGraph.hpp

```diff
@@ -65,90 +79,173 @@ class TGraph : public RGraph{
                 int dest;
                 int weight;
                 Relation(){}
-                Relation(Vertex* src, Vertex* dest, int weight): src(src->get_id()), dest(dest->get_id()), weight(weight){}
+                Relation(int src, int dest, int weight): src(src), dest(dest), weight(weight){}
                 Relation(const Relation& oth): src(oth.src), dest(oth.dest), weight(oth.weight){}
                 void set_weight(int w) {
                     weight = w;
                 }
                 int get_weight() {
                     return weight;
                 }
+
+                bool operator==(const Relation *other) const {
+                    return this->src == other->src && this->dest == other->dest;
+                }
         };
 
+        // Allocates data structures and pre-loads the graph
         TGraph(GlobalTestConfig* gtc) {
-            idxToVertex = new Vertex*[numVertices];
-            // Initialize...
-            for (size_t i = 0; i < numVertices; i++) {
-                idxToVertex[i] = new Vertex(i, -1);
+            srand(time(NULL));
+            size_t sz = numVertices;
+            this->idxToVertex = new Vertex*[numVertices];
+            std::cout << "Allocated idxToVertex..." << std::endl;
+            this->vertexLocks = new std::atomic<bool>[numVertices];
+            std::cout << "Allocated vertexLocks..." << std::endl;
+            this->vertexSeqs = new uint32_t[numVertices];
+            std::cout << "Allocated vertexSeqs..." << std::endl;
+            std::mt19937_64 gen(rand());
+            std::uniform_int_distribution<> verticesRNG(0, numVertices - 1);
+            std::uniform_int_distribution<> coinflipRNG(0, 100);
+            std::cout << "Allocated core..." << std::endl;
+            // Fill to vertexLoad
+            for (int i = 0; i < numVertices; i++) {
+                if (coinflipRNG(gen) <= vertexLoad) {
+                    idxToVertex[i] = new Vertex(i,i);
+                } else {
+                    idxToVertex[i] = nullptr;
+                }
+                vertexLocks[i] = false;
+                vertexSeqs[i] = 0;
+            }
+
+            std::cout << "Filled vertexLoad" << std::endl;
+
+            // Fill to mean edges per vertex
+            for (int i = 0; i < numVertices; i++) {
+                if (idxToVertex[i] == nullptr) continue;
+                for (int j = 0; j < meanEdgesPerVertex * 100 / vertexLoad; j++) {
+                    int k = verticesRNG(gen);
+                    while (k == i) {
+                        k = verticesRNG(gen);
+                    }
+                    if (idxToVertex[k] != nullptr) {
+                        Relation *in = new Relation(i, k, -1);
+                        Relation *out = new Relation(i, k, -1);
```

Comment:
> why do we create two relation to represent the same edge?
To prevent issues where I run into a segmentation fault in cases where I accidentally free a shared `Relation` early and to simplify the design (originally I used a `std::shared_ptr but removed it once was running into bugs and decided to take it one step at a time).

Reply:
An alternative could be always to free relations in src of vertices only, and use dest just as auxiliary indexing. Am I right?